### ISSUE TITLE
Refactor the ODE module using classes for solvers

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#18348]
This is a suitable GSOC idea.

Currently the ODE module (sympy/solvers/ode.py) which implements the `dsolve` function is messy. The code there would be easier to understand and maintain if each of the separate solvers was implemented as a class.

The idea here is to have classes like
```python
class Solver:
    def __init__(self, eq, func):
        # Store the inputs passed to dsolve
        self.eq = eq
        self.func = func

class FirstLinear(Solver):
    def match(self):
        # return True if this method can solve the equation

    def get_general_solution(self):
        # return the general solution
```

The `dsolve` can be rewritten so it looks more like this:
```python
all_solvers = [
    FirstLinear,
    ...
]

def dsolve(eq, func):
    solvers = []
    for solvercls in all_solvers:
        solver = solvercls(eq, func)
        if solver.matches():
            solvers.append(solver)

    # Get the first solver that matched and use it to return the solution
    return solvers[0].get_general_solution()
```
Rewriting the code like this will make it much easier to understand and maintain. This can be done one solver at a time so if interested in this as a GSOC idea you could start beforehand by doing one or two solvers.

The real benefit of doing this will come at the end when `dsolve` can be written like
```python
def dsolve(eq, func):
    solvers = []
    for solvercls in all_solvers:
        solver = solvercls(eq, func)
        if solver.matches():
            # Shortcut this loop and return the solution immediately
            return solver.get_general_solution()
```
This will make dsolve *much* faster for most ODEs.

**Comment by Mohitbalwani26** [comment]
so different methods should have different classes?

**Comment by risubaba** [comment]
@oscarbenjamin I would like to try doing this. However i am not really familiar with `ode.py`. Could you please guide me how i should start with this.

**Comment by risubaba** [comment]
In `_desolve()`
```diff
-    type = kwargs.get('type', None)
+    de_type = kwargs['type']
     xi = kwargs.get('xi')
     eta = kwargs.get('eta')
     x0 = kwargs.get('x0', 0)
     terms = kwargs.get('n')
 
-    if type == 'ode':
+    if de_type == 'ode':
         from sympy.solvers.ode import classify_ode, allhints
         classifier = classify_ode
         string = 'ODE '
         dummy = ''
 
-    elif type == 'pde':
+    else:
         from sympy.solvers.pde import classify_pde, allhints
         classifier = classify_pde
         string = 'PDE '

```
This should work as kwargs will always have a `type` key.

**Comment by oscarbenjamin** [comment]
> This should work as kwargs will always have a `type` key.

I'm not sure why you propose that change here. It might be a reasonable change but I think it is unrelated to the proposal here. You can open a new issue if you have an idea that you want to discuss.

> However i am not really familiar with `ode.py`. Could you please guide me how i should start with this.

I suggest reading the docs for `dsolve` and the ODE module. Then take a look at the code in the module and read the large comment at the top of the module that explains the current structure.

If you are interested in working on the ODE module then you should look at issues and PRs with the `solvers.dsolve` label:
https://github.com/sympy/sympy/labels/solvers.dsolve

**Comment by risubaba** [comment]
Thanks for the response. I suggested the initial change as not just the structure but code style of various portions can be changed as well. Should i raise a different issue for such style related queries?

**Comment by V1krant** [comment]
@oscarbenjamin, While writing the code for `class FirstLinear`, I noticed the following symbols are used extensively. 
https://github.com/sympy/sympy/blob/f5413ff89530b14ffc9b7359717f8c7ad03d2c6c/sympy/solvers/ode.py#L999-L1018
Should we declare these variables globally or only when required inside the `match` function, or are the two equally efficient?

**Comment by oscarbenjamin** [comment]
@V1krant I think that we should try to separate this out so we have functions that get what each solver needs. For example those wild symbols are used to get the coefficients of a linear ODE. We should make a function `_get_coeffs_linear_ode(eq, func, sym)` that does that and then use that in `FirstLinear`.

**Comment by oscarbenjamin** [comment]
@risubaba 
> Should i raise a different issue for such style related queries?

Yes if you want to discuss bigger changes.

For smaller changes just open a pull request. Then we can see if the tests pass.

**Comment by V1krant** [comment]
> We should make a function `_get_coeffs_linear_ode(eq, func, sym)` that does that and then use that in `FirstLinear`.

It sounds like a good idea. But I think that the same function (`_get_coeffs_linear_ode`) would be called again for calculating the solution in `get_general_solution`. Should we consider some workaround over this, or am I missing something?

**Comment by oscarbenjamin** [comment]
> But I think that the same function (`_get_coeffs_linear_ode`) would be called again for calculating the solution in `get_general_solution`

Don't worry about that now. Once the code is nicely organised it will be easy to make things more efficient.

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin should a pr be opened after implementing one or two solvers? Because if we do that than travis will obviously fail

**Comment by oscarbenjamin** [comment]
I don't see why Travis would need to fail. If you are interested in working on this then I suggest picking a particular solver (say which here so that others know) and then trying to factor it out into a class.

Different PRs will end up sharing some code so if there are several then one of them should be merged first and then the others will need to be made compatible with it afterwards.

**Comment by oscarbenjamin** [comment]
I realise that my description here was not very clear so I have opened a PR that does what I was thinking for the `nth_algebraic` solver (#18403). Any comments/reviews/questions on that PR are welcome.

**Comment by asmeurer** [comment]
Another thing we should think about is how to better handle meta-hints that manipulate the ODE somehow then fallback to other solvers to actually solve it. An example would be a meta-hint for solving a system with the strongly connected components https://github.com/sympy/sympy/issues/16174. Another example is the hint that reduces the order of an ODE that isn't a function of f(x), only its derivatives. Currently this hint is implemented, but it's not great because it doesn't (and can't) pass through any of the options to dsolve https://github.com/sympy/sympy/blob/35be3a042e38b1ef5de33120e38293540f40d5f5/sympy/solvers/ode/ode.py#L4204. Another meta-hint would be a whole class of solvers that apply a substitution to an ODE to transform it into a solvable type. 

One thing that would help is if the `hint` argument to `dsolve` could be some kind of hint class. Then for instance, one could create a generic "substitution" hint by manually specifying the substitution.  This would allow metahints to themselves include hints for which solvers to pass down to. Other hints may also find use from flags. Currently some solvers like the homogeneous order solver are split into two hints, but this could very well be a flag to a single hint. The existing string hints would also remain for backwards compatibility and simplicity.

**Comment by oscarbenjamin** [comment]
There should be a `desubs` function for doing substitutions into ODEs. That would be useful separately from the refactor here.

Solvers that work by substitution should probably be implemented by literally doing the substitution. Then the solution of the substituted ODE wouldn't need to be handled by the solver itself which could simply return the necessary substitution back to `dsolve/classify_ode`. That would make it possible for dsolve to accept hints that describe what to do after the substitution. More generally perhaps this could end up looking something like `integral_steps` which is explicitly organised into rules that lead into other rules etc.

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin For pattern matching, I kept in mind that we can extract the elements of our general solution from the equation with direct matching just like First_linear. And for `SingleODESolver` there will be proper logic checking whether the given equation matches or not.

I am a bit confused about how all linear solvers can be based on pattern because
let's say we want to implement `nth_linear_constant_coeff_undetermined_coefficients`.
its general equation is 

    a_n f^{(n)}(x) + a_{n-1} f^{(n-1)}(x) + .. + a_1 f'(x)  + a_0 f(x) = P(x)

Now p(x) needs to have a finite number of linearly independent derivatives and in pattern matching to write general solution we should use the extracted elements given by wilds function.

**Comment by oscarbenjamin** [comment]
Both `_wilds` and `_equation` take an order argument that can be used to determine the order of the ODE so e.g.:
```
In [6]: order = 4                                                                                                                                             

In [7]: derivs = [Derivative(f(x), (x, i)) for i in range(order)]                                                                                               

In [8]: derivs                                                                                                                                                
Out[8]: 
⎡                  2          3      ⎤
⎢      d          d          d       ⎥
⎢f(x), ──(f(x)), ───(f(x)), ───(f(x))⎥
⎢      dx          2          3      ⎥
⎣                dx         dx       ⎦

In [9]: P = Wild('P', exclude=[derivs]) 
```

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin sorry I should have explained more clearly. Actually what I want to say is in general equation p(x) should always consist of `sin,cos,x**n,exp(x),etc` because we need to check them for trialset.

In my opinion, through pattern matching, we try to extract elements through wilds function and we return an equation that should match the pattern in terms of these extracted elements.
 
Although I will try my best to refactor this with pattern matching.

**Comment by oscarbenjamin** [comment]
For the undetermined coefficients solver there would need to a be a separate `_verify` call as there is in `AlmostLinear`.

**Comment by suryam35** [comment]
@oscarbenjamin on which unmerged PR's should i start working with ?

**Comment by anirudhprabhakaran3** [comment]
Hello!

**Comment by anirudhprabhakaran3** [comment]
Hello!

I wish to work on this for GSoC. Are there any pre-requisites before making a proposal?
